### ISSUE TITLE
Define complete graph (how to build `edge_index` efficiently)

### ISSUE THREAD

**Issue by chi0tzp** [rusty1s/pytorch_geometric#964]
## ❓ Questions & Help

I would like to build a complete undirected graph, and I'm wondering if there is any built-in method for doing so. What really needs to be done here is the creation of the `edge_index`.

What I've done so far is shown in the following naive example. Here I do not include any self-loops, and another question is whether including self-loops in the `edge_index` would be correct.

```
import torch
from torch_geometric.data import Data as gData
from torch_geometric.utils import to_networkx
import networkx as nx
import matplotlib.pyplot as plt
import numpy as np

def build_edge_idx(num_nodes):
    # Initialize edge index matrix
    E = torch.zeros((2, num_nodes * (num_nodes - 1)), dtype=torch.long)
    
    # Populate 1st row
    for node in range(num_nodes):
        for neighbor in range(num_nodes - 1):
            E[0, node * (num_nodes - 1) + neighbor] = node

    # Populate 2nd row
    neighbors = []
    for node in range(num_nodes):
        neighbors.append(list(np.arange(node)) + list(np.arange(node+1, num_nodes)))
    E[1, :] = torch.Tensor([item for sublist in neighbors for item in sublist])
    
    return E
```

So, given the number of nodes, `num_nodes`, (and maybe other parameters, such as the number of node features, edge attributes, etc), I define my graph as follows (below I also use `networkx` for visualizing the graph):

```
num_nodes = 5
num_node_features = 16
num_edge_features = 1

x = torch.randn((num_nodes, num_node_features), dtype=torch.float)

edge_index = build_edge_idx(num_nodes=num_nodes)
num_edges = edge_index.size(1)
edge_attr = torch.randn(num_edges, num_edge_features)

data = gData(x=x, edge_index=edge_index, edge_attr=edge_attr)

G = to_networkx(data)

plt.plot()
nx.draw(G, with_labels=True)
```

My question is if there exists any built-in method in PyTorch Geometric for doing that efficiently.

**Comment by rusty1s** [comment]
There is no such functionality. The reason is that using PyG on fully-connected graphs is not particular useful, since it expects sparse connections. You can achieve way better performance by operating on dense adjacencies (or omit them completely).

**Comment by chi0tzp** [comment]
Hi @rusty1s, thanks for your quick response. Just a question: if we're talking about spatial convolutions on complete graphs, without pooling or up-scaling operations, `edge_index` should be defined just once, right?

**Comment by vedal** [comment]
@rusty1s when you say 
> You can achieve way better performance by operating on dense adjacencies (or omit them completely).
Do you mean to use dense adjacency matrix and for example a `DenseGCNConv` layer?

I'm trying to implement a network with learnable connectivity (edges are weights in the network) with a small graph (less than 10 nodes). Would you suggest that the fastest performance on this setup would be using `DenseGCNConv` and a dense adjacency? I haven't seen any docs using `DenseGCNConv`, even for small graphs like Karate dataset, so I was a bit confused. 

Thanks in advance, and also thanks alot for this great library! :)

**Comment by rusty1s** [comment]
Yes, we are supporting a few dense GNN operators by now, which should be pretty fast if your graph is inherently dense. With my previous reply, I just wanted to say that it is not recommended to use the sparse `edge_index` format when operating on fully-connected graphs.

**Comment by vedal** [comment]
@rusty1s I understand! Thanks alot for the swift reply :)

**Comment by andrei-rusu** [comment]
Relating to this issue, it seems as per https://openreview.net/forum?id=i80OPhOCVH2 that adding a "fully-adjacent" layer to a GNN improves its capability to capture long-range dependencies (i.e. by offsetting the so-called 'bottleneck'). 

I've been trying to add such a layer to my architecture by appending a final DenseGCNConv layer, which gets fed with the transformed data, a fully-connected adjacency matrix, and no mask (in the transductive setting). However, it seems that the model is unable to learn anything on Cora/Citeseer. I got a bit more lucky with DenseSAGEConv and DenseGraphConv (it looks like the 'mean' aggr allows for some signal to be correctly backprop'ed). Even still, the performance is worse than having a simple graph conv layer with normal inputs.

My question is: How would you go about adding such a 'complete-graph' layer to an architecture? And if I am on the right track, do you have any idea as to why the model may not be able to learn?

Thank you so much for the library, btw. Amazing work!

**Comment by rusty1s** [comment]
Yes, the `DenseGCNConv` layer does not really work on a fully-connected graph, as it will produce an equal embedding for all nodes. This is avoided in the `DenseSAGEConv` layer as it will maintain the original node features, and simply adds the mean representation of all nodes into its representation. Instead of using pre-defined layers, you can also write that quite easily by yourself:
```python
x = conv(x, edge_index)
global_x = self.linear(global_mean_pool(x, batch))
x = x + global_x[batch]
```

**Comment by andrei-rusu** [comment]
Thank you very much for taking your time to answer this! Cheers.

**Comment by gorkamunoz** [comment]
> Yes, the `DenseGCNConv` layer does not really work on a fully-connected graph, as it will produce an equal embedding for all nodes. 

Hi @rusty1s, I am seeing this effect happening when applying GNN layers to a fully connected graph (both with `GCNConv` and `GATv2Conv`). Could you give me any insight on why is it happening? From the equations of the layers, I was not expecting such behaviour.. Thanks!

**Comment by gorkamunoz** [comment]
Actually, I see this happening with `GATvsConv` only if I consider undirected graphs. Also I have to say that it does happen consistenly, and sometimes I get identical outputs and sometimes different ones. As you said, if I look at the attention weights, I see that all have the same value. This is an example, for which I define:

```python
gat  = GATv2Conv(8,1,heads = 1)

x = torch.rand(4, 8)
```

1) **Directed graph:** Everything looks normal here.
![image](https://user-images.githubusercontent.com/25931612/226836847-8fe9f0eb-8756-45b7-8577-ea2faf8d3f0f.png)


2) **Undirected graph:** Rather consistently, attention weights are equal between a given $i$ and all its $j$ (following the notation from the [doc page](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.GATv2Conv.html). This moreover makes all the outputs equal...
![image](https://user-images.githubusercontent.com/25931612/226837323-8f2bbf4f-03fc-4ed7-b76c-ce6022c41f88.png)

Looking at the equations, I really don't see why $\alpha_{i,j}$ is equal for any $j$, as at least $\Theta\[x_i || x_j\]$ should vary with $j$. 

As said, this effect does not happen consistently and sometimes I actually get different output values and different attention weights. However, I can't find what is causing this to happen, see for instance:

![image](https://user-images.githubusercontent.com/25931612/226839148-81f7f21d-5c0d-45ba-8448-fda4b35cb196.png)

Thanks a lot for your help!

**Comment by rusty1s** [comment]
Ah, you are using `GATv2Conv`, sorry I misread. As far as I see, `GATv2Conv` uses a symmetric an symmetric attention mechanism, so this is indeed as well expected for `GATv2Conv`.

**Comment by rusty1s** [comment]
Mh, I think you are correct, but I am not totally sure why TBH. It looks like attention scores across different neighborhoods are just translated, and the `softmax` will thus produce equal attention weights afterwards. Very interesting finding.

**Comment by RayZ0rr** [comment]
@rusty1s  sorry bumping this issue but [`DenseSAGEConv`](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.dense.DenseSAGEConv.html#torch_geometric.nn.dense.DenseSAGEConv) expects the adajceny matrix as its second argument.

How does it work for graphs with different number of nodes? Does the graph require padding so that all graphs have the same number of nodes?

I can construct a fully-connected adjacency matrix for one graph with :
```
adj_matrix = torch.ones((num_nodes, num_nodes)) - torch.eye(num_nodes)
```
But this will only apply to the graph with `num_nodes` number of nodes. How to apply it in batch?

**Comment by rusty1s** [comment]
Yes, this will require padding, which `to_dense_batch` and `to_dense_adj` will take care of :)

**Comment by RayZ0rr** [comment]
@rusty1s  Thanks for the quick reply. Just to confirm before implementing, I will have to construct `edge_index` for fully constructed graph as mentioned in the First post and pass it to `to_dense_adj`. Also pass `x` to `to_dense` and finally take the outputs from these to `DenseSAGEConv` class operator.

**Comment by Maximilian-Stefan-Ernst** [comment]
@rusty1s, @gorkamunoz did you figure out why the `GATv2Conv` layer produces equal embeddings for fully connected graphs? I got the same problem, and could not find anything about this behaviour except this issue...

This is especially weird as graph attention networks are recommended for the setting of fully connected graphs?
